pkg/openstack/actuator.go

```diff
@@ -287,5 +288,15 @@ func (a *OpenStackActuator) getLogger(cr *minterv1.CredentialsRequest) log.Field
 // if the system is considered not upgradeable. Otherwise, return nil as the default
 // value is for things to be upgradeable.
 func (a *OpenStackActuator) Upgradeable(mode operatorv1.CloudCredentialsMode) *configv1.ClusterOperatorStatusCondition {
-	return nil
+	upgradeParams := &utils.UpgradeableParams{
+		Mode:               mode,
+		TargetRelease:      constants.UpgradeableNextRelease,
+		ExpectedNewSecrets: a.getUpcomingCredSecrets(),
+		RootSecretName:     a.GetCredentialsRootSecretLocation(),
+	}
+	return utils.UpgradeableCheck(a.Client, upgradeParams)
```

Comment:
Perhaps the actuators should return a list of problematic versions, and if none are set, the annotation is required. For example on AWS we need to do this, on other clouds we do not presently, so it would be a shame to make them do this work for no reason, especially in the absence of a "manual mode cred reconcile tool". (btw I really suggest keeping that in mind when looking into that STS tooling)

Reply:
The sts-preflight tool kinda-sorta does this (or if it doesn't it can easily be extended to do this). You fetch the CredentialsRequest list from whatever version of OpenShift you're interested in, and the sts-preflight tool will update any existing Roles if they need new permissions. It also creates the Secret yaml, but doesn't presently automatically apply the Secrets.

Reply:
And does it do this in an additive fashion, i.e. it's ok to make new creds, or add permissions, but do not reconcile removal of any creds or permissions because i'm probably not on that version yet? IMO the tool for this purpose should have a mode/flag to opt-in to "reconcile and allow removing creds/perms" vs additive only preupgrade mode (which is probably the default)

Reply:
sts-preflight has no idea about CredRequest that you do not provide to it. It just converts a list of CredRequest yaml to IAM Roles and kubernetes Secrets (with predictable names). I suppose you could provide it CredRequests with non-nil deletionTimestamp to trigger a deletion, but it wasn't something that has been considered up to this point.
Maybe @akhil-rane 's cleanup work can annotate CredRequest that this tool can see and take action with.